hazelcast/src/main/java/com/hazelcast/map/eviction/CustomEvictionPolicy.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.eviction;
+
+import com.hazelcast.core.EntryView;
+
+import java.util.Comparator;
+
+/**
+ * An eviction policy takes its role after the decision that a map needs to free some memory
+ * according to one of {@link com.hazelcast.config.MaxSizeConfig.MaxSizePolicy MaxSizePolicies}
+ * and it helps to find most appropriate entries to remove.
+ *
+ * {@link com.hazelcast.core.IMap IMap} has out-of-the-box eviction policies like {@link LRUEvictionPolicy},
+ * {@link LFUEvictionPolicy} and {@link RandomEvictionPolicy} but if there is a need to define another eviction policy,
+ * users can introduce a custom one by extending {@link CustomEvictionPolicy} class.
+ *
+ * <p>Implementation:</p>
+ * {@link CustomEvictionPolicy#compare(EntryView, EntryView)} method must be implemented to provide an ascending order
+ * of entries. Because internal eviction algorithm will evict the smallest entry from {@link com.hazelcast.core.IMap IMap}
+ *
+ * @param <K> the type of keys maintained by IMap
+ * @param <V> the type of mapped values
+ *
+ * @see LRUEvictionPolicy
+ * @see LFUEvictionPolicy
+ * @see RandomEvictionPolicy
+ *
+ * @since 3.7
+ *
+ */
+public abstract class CustomEvictionPolicy<K, V> implements Comparator<EntryView<K, V>> {
```

Comment:
@serkan-ozal: I understand your objection. I don't think I can make a general rule for all cases. I'd certainly never say "use abstract classes everywhere". It has to be considered on case-by-cases basis. 

When it comes to extension point for users then I'd say intefaces make sense in cases where the contract is not the main concern of implementing classes. Various `FooAware` or perhaps even `BarListener`.

However I'd say in the case of eviciton stratgy an abstract class makes sense. It makes API evolution straightforward. Imagine we'll have this requirement: "I want my strategy to be notified when a new entry is added to a map". It's pretty simple when you have an abstract class. 

The alternative approach is to have 1 method per interface. Then as you are evolving your code  is either cluttered with
``java
if (foo instaceof WeatherAware) {
  WeatherAware wa = (WeatherAware)foo;
  wa.stormStarted();
}
```
or you need to use adapters internally.

It's not uncommon to provide abstract classes as extension points in other projects too. This is quite well-known example: https://github.com/hamcrest/JavaHamcrest/blob/f681d7e29265ad8aed7b46e7d9737c497ad2fd1f/hamcrest-core/src/main/java/org/hamcrest/Matcher.java#L60-L60 :)